resources/prosody-plugins/mod_visitors_component.lua

```diff
@@ -125,25 +130,36 @@ local function stanza_handler(event)
         return;
     end
 
+    local processed;
+    -- promotion request is coming from visitors and is a set and is over the s2s connection
     local request_promotion = visitors_iq:get_child('promotion-request');
     if request_promotion then
         respond_iq_result(origin, stanza);
-        return request_promotion_received(room, request_promotion.attr.jid, stanza.attr.from);
+        processed = request_promotion_received(room, request_promotion.attr.jid, stanza.attr.from);
     end
 
+    local is_from_admin = is_admin(jid.bare(stanza.attr.from));
+
+    -- this is only from jicofo, so it is safe to
     local connect_vnode = visitors_iq:get_child('connect-vnode');
-    if connect_vnode then
+    if connect_vnode and is_from_admin then
         respond_iq_result(origin, stanza);
-        return connect_vnode_received(room, connect_vnode.attr.vnode);
+        connect_vnode_received(room, connect_vnode.attr.vnode);
+        processed = true;
     end
 
     local disconnect_vnode = visitors_iq:get_child('disconnect-vnode');
-    if disconnect_vnode then
+    if disconnect_vnode and is_from_admin then
```

Comment:
What about support for multiple elements with the same type? Jicofo sends multiple `<disconnect-vnode>`s

Reply:
Ups.